### ISSUE TITLE
Cannot link for musl under x86_64 via rust

### ISSUE THREAD

**Issue by mqudsi** [rui314/mold#330]
Attempting to build any rust project with `cargo build --release --target x86_64-unknown-linux-musl` with `rustc` configured to use `mold` via the environment variable `RUSTFLAGS=-C link-arg=-B/opt/mold/` with `/opt/mold/ld` being a symlink to `/usr/bin/mold` under Linux (Ubuntu 20.04) with `cc` pointing to GCC 11.1 causes compilation to fail with the following error:

```
note: "cc" "-m64" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/rcrt1.o" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crti.o" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtbeginS.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.0.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.1.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.10.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.11.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.12.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.13.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.14.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.15.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.2.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.3.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.4.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.5.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.6.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.7.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.8.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.83c7e763-cgu.9.rcgu.o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.28d512yoe8rf5bry.rcgu.o" "-Wl,--as-needed" "-L" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps" "-L" "/home/mqudsi/hddrand/target/release/deps" "-L" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib" "-Wl,-Bstatic" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libsize-df1de5e4d09f8d5b.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libnum_traits-41c13dbcec65677d.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/librand-5fcb6e742b92f4a5.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/librand_chacha-fbfd9fb78a768d1d.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libppv_lite86-70fc9d59b0bca089.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/librand_core-b2a1f16fc65e42db.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libgetrandom-ab8f293045b793fd.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/liblibc-d0b027d431f4ce47.rlib" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libcfg_if-1bc3169da7929e1d.rlib" "-Wl,--start-group" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-46841458d5b53520.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libpanic_unwind-8e3ab72bb78c6ae7.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libminiz_oxide-320620266971f437.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libadler-e629c7f08b6e578d.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libobject-0ba935e4cbaf9b4e.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libmemchr-f80756333d5fa71f.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libaddr2line-6ac1f1426f978e95.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libgimli-9ca1fee54ff3f19b.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd_detect-501dbaeebe011100.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_demangle-ce3a8a871a335a27.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libhashbrown-b74ac84e30e2374e.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_alloc-23eed3b2a2a2cef5.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libunwind-13d48a01dbbb98e5.rlib" "-lunwind" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcfg_if-16e1723e405e8810.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liblibc-e25734a7bb04e2b6.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liballoc-6d66cb1e305bead6.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_core-d23d1fdf2f9c2e82.rlib" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcore-4927c240c984c2ab.rlib" "-Wl,--end-group" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcompiler_builtins-d864e382333f88f0.rlib" "-Wl,-Bdynamic" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-nostartfiles" "-L" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib" "-L" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained" "-o" "/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15" "-Wl,--gc-sections" "-static-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "-B/opt/mold/" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtendS.o" "/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtn.o"
  = note: mold: cannot open text
          collect2: error: ld returned 1 exit status
          

error: could not compile `hddrand` due to previous error
```

I found a similar issue that was closed because it could not be reproduced with open source code. Perhaps this will help:

This was tested with the latest mold `master` after `cd`ing into a checkout of https://github.com/mqudsi/hddrand and executing `cargo build --release --target x86_64-unknown-linux-musl`. Using the latest version of the nightly rust compiler with the x86_64-unknown-linux-musl toolchain installed.

Clearing `RUSTFLAGS` before compiling (i.e. no longer using mold) causes the build to complete OK.

**Comment by rui314** [comment]
My `cc` is GCC 10.3, and `hddrand` builds fine on my machine.

```
ruiu@blue:~/hddrand$ cc --version
cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

ruiu@blue:~/hddrand$ RUSTFLAGS='-C link-arg=-B/usr/local/libexec/mold' cargo build --release --target x86_64-unknown-linux-musl
   Compiling libc v0.2.68
   Compiling cfg-if v1.0.0
   Compiling autocfg v1.0.0
   Compiling ppv-lite86 v0.2.16
   Compiling num-traits v0.2.11
   Compiling getrandom v0.2.4
   Compiling rand_core v0.6.3
   Compiling rand_chacha v0.3.1
   Compiling rand v0.8.4
   Compiling size v0.1.2
   Compiling hddrand v0.1.1 (/home/ruiu/hddrand)
    Finished release [optimized] target(s) in 2.55s
```

My distro is Ubuntu 20.04. It looks like GCC 11.1 is not available via `apt`. Did you build gcc from source?

It looks like mold got an unknown option `text` as a command line option. I wonder if there's a way to see the actual command line options passed to `mold`.

**Comment by mqudsi** [comment]
I installed gcc 11 via the ubuntu-toolchain-r ppa at https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/ppa (it seems it now has gcc 12 as well!) - but that doesn't seem to be related.

It seems the problem takes place when `/opt/mold/ld` is a symlink to `mold`. I discovered this by trying to answer your question

> I wonder if there's a way to see the actual command line options passed to `mold`.

So I replaced `/opt/mold/ld` (changing it from a symlink to an executable) with the following script:

```sh
#!/bin/sh

echo "$@" > /tmp/mold.txt
exec /usr/bin/mold "$@"
```

Which gave me the invocation parameters (attached for posterity: [mold.txt](https://github.com/rui314/mold/files/8010576/mold.txt)), **but built the executable just fine**.

I'm not sure why this is only exhibiting when building for a non-default (or just musl) target but not in the normal case. When omitting `--target ...`, `mold` is invoked with the following command line:
[mold (default).txt](https://github.com/rui314/mold/files/8010586/mold.default.txt)

The diff of the two (a is musl which fails via symlink, b is default target which doesn't), with the spaces replaced by new lines:

```diff
4c4
< -plugin-opt=-fresolution=/tmp/ccViN3Ds.res
---
> -plugin-opt=-fresolution=/tmp/ccwBGkpY.res
11,15c11,12
< -static
< -pie
< --no-dynamic-linker
< -z
< text
---
> -dynamic-linker
> /lib64/ld-linux-x86-64.so.2
22,23c19,22
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15
< -L/home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps
---
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5
> /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/Scrt1.o
> /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crti.o
> /usr/lib/gcc/x86_64-linux-gnu/11/crtbeginS.o
25,27c24,25
< -L/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib
< -L/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib
< -L/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained
---
> -L/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib
> -L/home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib
37,56c35,51
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/rcrt1.o
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crti.o
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtbeginS.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.0.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.1.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.10.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.11.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.12.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.13.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.14.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.15.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.2.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.3.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.4.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.5.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.6.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.7.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.8.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.hddrand.f6b7c93b-cgu.9.rcgu.o
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/hddrand-c462964cd557ac15.1g5oj8t2r1pmdg4k.rcgu.o
---
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.0.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.1.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.10.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.11.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.12.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.13.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.14.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.15.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.2.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.3.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.4.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.5.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.6.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.7.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.8.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.hddrand.9131cb12-cgu.9.rcgu.o
> /home/mqudsi/hddrand/target/release/deps/hddrand-af7f9d878f789cd5.3qib4uheo00go4ul.rcgu.o
59,67c54,62
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libsize-df1de5e4d09f8d5b.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libnum_traits-41c13dbcec65677d.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/librand-5fcb6e742b92f4a5.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/librand_chacha-fbfd9fb78a768d1d.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libppv_lite86-70fc9d59b0bca089.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/librand_core-b2a1f16fc65e42db.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libgetrandom-ab8f293045b793fd.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/liblibc-d0b027d431f4ce47.rlib
< /home/mqudsi/hddrand/target/x86_64-unknown-linux-musl/release/deps/libcfg_if-1bc3169da7929e1d.rlib
---
> /home/mqudsi/hddrand/target/release/deps/libsize-10be0721cb9402d9.rlib
> /home/mqudsi/hddrand/target/release/deps/libnum_traits-80615104634bff22.rlib
> /home/mqudsi/hddrand/target/release/deps/librand-bad5a60a4895acce.rlib
> /home/mqudsi/hddrand/target/release/deps/librand_chacha-4294b984cf72f41e.rlib
> /home/mqudsi/hddrand/target/release/deps/libppv_lite86-bf7dadb5db4d146d.rlib
> /home/mqudsi/hddrand/target/release/deps/librand_core-3c1c85845318c254.rlib
> /home/mqudsi/hddrand/target/release/deps/libgetrandom-a294a2a747698520.rlib
> /home/mqudsi/hddrand/target/release/deps/liblibc-5b51392b76c9d670.rlib
> /home/mqudsi/hddrand/target/release/deps/libcfg_if-af4ed11818f18246.rlib
69,88c64,81
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-d4732579dd1df351.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libpanic_unwind-a02ea0a122bc57d6.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libminiz_oxide-eb774c0d299187b5.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libadler-ab00a30bdea74bfd.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libobject-f0afe04e837001a3.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libmemchr-e63707400f7f6e4d.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libaddr2line-1c2b0357e1a7fb61.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libgimli-0187031323e333d5.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_demangle-9d63941f8a36d464.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd_detect-9e73f7a2f4e541a7.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libhashbrown-46a56239e2704caa.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_alloc-c06d1dcb5fe59062.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libunwind-33cff1364a61948e.rlib
< -lunwind
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcfg_if-f3aade1f3a3d93e2.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liblibc-c4a07f47e2b60107.rlib
< -lc
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liballoc-b8ec8a58684bdf85.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_core-d23d1fdf2f9c2e82.rlib
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcore-4927c240c984c2ab.rlib
---
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-15eca284732d16a5.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-2dea7b7b5d7d34da.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-e59f5bd4cb2cfaf0.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-7668ff0023878c7d.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-458744404e1def67.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-4efcda9b1cf322d9.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-37d6b06fd67de855.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-b86752c62a2c4219.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-18fdb65d66742a3a.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-7052aa9425c606b1.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-dcaf899f38c80f9e.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-29a7296fb6c88181.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-2c9f98aff68910ce.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-109657cf0c701d74.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-e869a247063ed69a.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-013f8207d93adfc9.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-2a6a2797f7a73818.rlib
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0e3656b1fda5fd7b.rlib
90c83
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcompiler_builtins-495e172e74f31dab.rlib
---
> /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-eb99f4fc7650a33a.rlib
91a85,98
> -lc
> -lm
> -lrt
> -lpthread
> -lutil
> -ldl
> -lutil
> -lgcc_s
> -lutil
> -lrt
> -lpthread
> -lm
> -ldl
> -lc
98,99c105,106
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtendS.o
< /home/mqudsi/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtn.o
---
> /usr/lib/gcc/x86_64-linux-gnu/11/crtendS.o
> /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/crtn.o
```

You can see very clearly how `< text` is no longer emitted when the static musl target is no longer selected.

**Comment by rui314** [comment]
I don't know why rustc passed different command line options to the shell script and the mold command, but this must be a bug fixed by 68fa44d39940bd2b5d44606d1bffd80cafa44fe1. Updating your mold binary should fix the issue.

**Comment by mqudsi** [comment]
Sorry, I may not have been very clear. rustc passes the same arguments regardless of whether `/opt/mold/ld` is a symlink or a script (different arguments only get passed when building for musl vs libc). When `/opt/mold/ld` is a script, mold handles `< text` fine. When `/opt/mold/ld` is a symlink, mold chokes on those same arguments (or that same `< text` argument in particular).

I had already updated to the latest `master` before filing the bug. I just updated and rebuilt again, and confirmed the bug persists:

```
$ mold --version
mold 1.0.3 (27c50e8158322932b62eb23771863ab3de40d5f4; compatible with GNU ld)
```
